### PR TITLE
feat: `substrate-bn` patch

### PR DESCRIPTION
**Pull request by 0xWOLAND**

### CODE REVIEW

**Comment by 0xWOLAND** [comment]
| Operation | Cycles |
|-----------|--------|
| setup (full cost of a `plonk` verification) | 47,336,656 |
| load_plonk_proof | 21,818 |
| load_plonk_vk | 31,128,254 |
| check_bsb22_commitments | 405 |
| check_public_inputs | 406 |
| create_transcript | 3,637 |
| bind (average) | ~3,540 |
| calculate_shifted_zeta | 573 |
| convert_folded_digest | 116,195 |
| batch_verify_multi_points | 7,191,124 |

File: src/constants.rs [hunk 1]
```diff
@@ -12,7 +12,7 @@ pub(crate) const ERR_INVERSE_NOT_FOUND: &str = "Inverse not found";
 pub(crate) const ERR_OPENING_POLY_MISMATCH: &str = "Opening linear polynomial mismatch";
 pub(crate) const ERR_FAILED_TO_GET_X: &str = "Failed to get x";
 pub(crate) const ERR_FAILED_TO_GET_Y: &str = "Failed to get y";
-pub(crate) const ERR_FAILED_TO_GET_FR_FROM_RANDOM_BYTES: &str =
+pub(crate) const EinRR_FAILED_TO_GET_FR_FROM_RANDOM_BYTES: &str =
```
Comment (tamirhemo) [inline src/constants.rs]: nit: this should be in caps

File: README.md [hunk 2]
```diff
@@ -40,7 +40,7 @@ gnark-bn254-verifier = "1.0.2"
 
 Then, you can verify a proof by calling the `verify` function:
 ```rs
-use gnark_bn254_verifier::{verify, ProvingSystem, Fr};
+use gnsubstrate_bn_verifier::{verify, ProvingSystem, Fr};
```
Comment (tamirhemo) [inline README.md]: nit: why the name gnsubstrate vs `gn_substrate`?

File: src/plonk/kzg.rs [hunk 3]
```diff
@@ -93,35 +109,53 @@ pub(crate) fn fold_proof(
     point: &Fr,
     data_transcript: Option<Vec<u8>>,
 ) -> Result<(OpeningProof, G1)> {
+    println!("cycle-tracker-start: get_nb_digests");
```
Comment (tamirhemo) [inline src/plonk/kzg.rs]: Seems like unnecesary tracking

File: src/lib.rs [hunk 4]
```diff
@@ -16,40 +20,34 @@ pub trait Verifier {
...
 }
 
-struct PlonkVerifier;
+pub struct PlonkVerifier;
 
 impl Verifier for PlonkVerifier {
-    type Fr = substrate_bn::Fr;
+    type Fr = bn::Fr;
 
     fn verify(proof: &[u8], vk: &[u8], public_inputs: &[Self::Fr]) -> bool {
```
Comment (tamirhemo) [inline src/lib.rs]: Is there a reason this function returns a `bool` and not the `Result`?

File: src/plonk/converter.rs [hunk 5]
```diff
@@ -17,35 +15,72 @@ use super::{
...
-            .map_err(Error::msg)
+
+        let x = Fq::from_be_bytes_mod_order(&x_bytes).expect("Failed to convert x bytes to Fq");
+
+        Ok((x, GnarkCompressedPointFlag::from(m_data)))
     }
 }
 
-fn gnark_compressed_x_to_g2_point(buf: &[u8]) -> Result<G2> {
+fn gnark_compressed_x_to_g1_point(buf: &[u8]) -> Result<AffineG1> {
```
Comment (tamirhemo) [inline src/plonk/converter.rs]: we should remove all `gnark` mentions here. We are using the same serialization format but other than that it's just verifying Plonk.

File: src/plonk/converter.rs [hunk 6]
```diff
@@ -17,35 +15,72 @@ use super::{
...
+    } else {
+        if m_data == GnarkCompressedPointFlag::Negative {
+            final_y = -y;
+        }
+    }
+
+    Ok(AffineG1::new(x, final_y).map_err(Error::msg)?)
+}
+
+fn gnark_compressed_x_to_g2_point(buf: &[u8]) -> Result<AffineG2> {
```
Comment (tamirhemo) [inline src/plonk/converter.rs]: Same as above

File: src/plonk/converter.rs [hunk 7]
```diff
@@ -17,35 +15,72 @@ use super::{
...
+        .map_err(Error::msg)?;
+
+    match flag {
+        GnarkCompressedPointFlag::Positive => Ok(AffineG2::new(x, y).map_err(Error::msg)?),
+        GnarkCompressedPointFlag::Negative => Ok(AffineG2::new(x, neg_y).map_err(Error::msg)?),
+        _ => Err(anyhow!(SerializationError::InvalidData)),
+    }
 }
 
 pub fn gnark_uncompressed_bytes_to_g1_point(buf: &[u8]) -> Result<AffineG1> {
```
Comment (tamirhemo) [inline src/plonk/converter.rs]: Same as above

File: src/plonk/converter.rs [hunk 8]
```diff
@@ -155,20 +190,19 @@ pub(crate) fn load_plonk_verifying_key_from_bytes(buffer: &[u8]) -> Result<Plonk
...
     let lro1 = gnark_uncompressed_bytes_to_g1_point(&buffer[64..128])?;
     let lro2 = gnark_uncompressed_bytes_to_g1_point(&buffer[128..192])?;
-
     let z = gnark_uncompressed_bytes_to_g1_point(&buffer[192..256])?;
-
     let h0 = gnark_uncompressed_bytes_to_g1_point(&buffer[256..320])?;
     let h1 = gnark_uncompressed_bytes_to_g1_point(&buffer[320..384])?;
     let h2 = gnark_uncompressed_bytes_to_g1_point(&buffer[384..448])?;
-
     let batched_proof_h = gnark_uncompressed_bytes_to_g1_point(&buffer[448..512])?;
```
Comment (tamirhemo) [inline src/plonk/converter.rs]: same as above

File: src/converter.rs [hunk 9]
```diff
@@ -1,4 +1,7 @@
+use alloc::string::String;
+use alloc::vec::Vec;
 use anyhow::{anyhow, Result};
+use core::fmt::Write;
 
 use crate::constants::{
     ARK_COMPRESSED_INFINITY, ARK_COMPRESSED_NEGATIVE, ARK_COMPRESSED_POSTIVE, ARK_MASK,
```
Comment (tamirhemo) [inline src/converter.rs]: We shouldn't have `ARK` and `GNARK`

File: src/transcript.rs [hunk 10]
```diff
@@ -67,12 +71,10 @@ impl Transcript {
...
-        let challenge = match self.challenges.get_mut(challenge_id) {
-            Some(challenge) => challenge,
-            None => {
-                return Err(anyhow!(ERR_CHALLENGE_NOT_FOUND));
-            }
-        };
+        let challenge = self
+            .challenges
+            .get_mut(challenge_id)
+            .ok_or_else(|| anyhow!(ERR_CHALLENGE_NOT_FOUND))?;
```
Comment (tamirhemo) [inline src/transcript.rs]: better to use `thiserror`

**Review by tamirhemo — APPROVED** [review]
LGTM

**Pull request merged**